Lesson4_wordbreak.cpp

```diff
@@ -0,0 +1,136 @@
+/*Word Break
+Given an input string and a dictionary of words, find out if the input string can be segmented into a space-separated sequence of dictionary words.
+For example, consider the following dictionary: { pear, salmon, foot, prints, footprints, leave, you, sun, girl, enjoy },
+
+Examples:
+Given the string “youenjoy”, 
+Output: True (The string can be segmented as “you enjoy”)
+
+Input: “youleavefootprints”,
+Output: True (The string can be segmented as “you leave footprints” or “you leave foot prints”)
+
+Input:salmonenjoyapples
+Output: False
+*/
+#include<iostream>
+#include<cstring>
+#include<vector>
+using namespace std;
+bool findword(string word,vector<string> dict)
+{
+	for(int i=0;i<dict.size();++i)
+	{
+		if(!(dict[i].compare(word)))
+			return true;
+	}
+	return false;
+	
+}
+bool wordbreak(string str,vector<string> dict)
+{
+	int strlen,dictlen;
+	strlen=str.size();
+	dictlen=dict.size();
+	if(strlen==0)
+		return true;
+	if(dictlen==0)
+		return false;
+	int dp[strlen+1]={0};
+	
+	//An array to store the results of the subproblem
+	for(int i=1;i<=strlen;++i)
+	{
+		if((dp[i]==0) && (findword(str.substr(0,i),dict)))
+		{
+			dp[i]=1;
+			
+		}
+		if(dp[i]==1)
+		{
+			if(i==(strlen))
```

Comment:
I would move this outside of the for loop since it happens at the end of the last iteration.  Then, after the loop you can `return dp[strlen]`